core/trie/src/node_codec.rs

```diff
@@ -92,53 +115,137 @@ impl<H: Hasher> trie_db::NodeCodec<H> for NodeCodec<H> {
+	fn leaf_node(partial: Partial, value: &[u8]) -> Vec<u8> {
+		let mut output = partial_enc::<N>(partial, NodeKind::Leaf);
 		value.encode_to(&mut output);
 		output
 	}
 
-	// FIXME: refactor this so that `partial` isn't already encoded with HPE. Should just be an `impl Iterator<Item=u8>`.
-	fn ext_node(partial: &[u8], child: ChildReference<H::Out>) -> Vec<u8> {
-		let mut output = partial_to_key(partial, EXTENSION_NODE_OFFSET, EXTENSION_NODE_BIG);
-		match child {
-			ChildReference::Hash(h) => 
-				h.as_ref().encode_to(&mut output),
-			ChildReference::Inline(inline_data, len) =>
-				(&AsRef::<[u8]>::as_ref(&inline_data)[..len]).encode_to(&mut output),
-		};
-		output
+	fn ext_node(
+		_partial: impl Iterator<Item = u8>,
+		_nbnibble: usize,
+		_child: ChildReference<<H as Hasher>::Out>,
+	) -> Vec<u8> {
+		unreachable!()
+	}
+
+	fn branch_node(
+		_children: impl Iterator<Item = impl Borrow<Option<ChildReference<<H as Hasher>::Out>>>>,
+		_maybe_value: Option<&[u8]>) -> Vec<u8> {
+		unreachable!()
 	}
 
-	fn branch_node<I>(children: I, maybe_value: Option<DBValue>) -> Vec<u8>
-		where I: IntoIterator<Item=Option<ChildReference<H::Out>>> + Iterator<Item=Option<ChildReference<H::Out>>>
-	{
-		let mut output = [0, 0, 0].to_vec();
-		let have_value = if let Some(value) = maybe_value {
-			(&*value).encode_to(&mut output);
-			true
+	fn branch_node_nibbled(
+		partial: impl Iterator<Item = u8>,
+		nb_nibble: usize,
+		children: impl Iterator<Item = impl Borrow<Option<ChildReference<<H as Hasher>::Out>>>>,
+		maybe_value: Option<&[u8]>) -> Vec<u8> {
+		let mut output = if maybe_value.is_some() {
+			partial_enc_it::<N,_>(partial, nb_nibble, NodeKind::BranchWithValue)
 		} else {
-			false
+			partial_enc_it::<N,_>(partial, nb_nibble, NodeKind::BranchNoValue)
+		};
+		let bm_ix = output.len();
+		let mut bm: BM::Buff = Default::default();
+		(0..BM::ENCODED_LEN).for_each(|_|output.push(0));
+		if let Some(value) = maybe_value {
+			value.encode_to(&mut output);
 		};
-		let prefix = branch_node(have_value, children.map(|maybe_child| match maybe_child {
+		BM::encode(children.map(|maybe_child| match maybe_child.borrow() {
 			Some(ChildReference::Hash(h)) => {
 				h.as_ref().encode_to(&mut output);
 				true
 			}
-			Some(ChildReference::Inline(inline_data, len)) => {
-				(&AsRef::<[u8]>::as_ref(&inline_data)[..len]).encode_to(&mut output);
+			&Some(ChildReference::Inline(inline_data, len)) => {
+				inline_data.as_ref()[..len].encode_to(&mut output);
 				true
 			}
 			None => false,
-		}));
-		output[0..3].copy_from_slice(&prefix[..]);
+		}), bm.as_mut());
+		output[bm_ix..bm_ix + BM::ENCODED_LEN].copy_from_slice(&bm.as_ref()[..BM::ENCODED_LEN]);
 		output
 	}
+
+}
+
+// utils
+
+/// Encode and allocate node type header (type and size), and partial value.
+/// It uses an iterator over encoded partial bytes as input.
+fn partial_enc_it<N: NibbleOps, I: Iterator<Item = u8>>(
```

Comment:
avoid use of arbitrarily placed shorted words in naming. it just looks horrible and makes everything unreadable and incomprehensible. if you mean "iterator", use either `_iterator` or, since we already have the convention, `_iter`. if you mean "encrypted", use `_encrypted`. if you mean "encoded", then use `_encoded`. those extra characters can mean the difference between taking 1 second to understand and taking 1 minute to understand.